File: leetcode_c_plus_plus/386.cpp
//
//
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 386.cpp
@Time : 2022/04/18 20:04:10
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
**/
class Solution {
public:
    vector<int> lexicalOrder(int n) {
        vector<int> ans(n);
        int number = 1;
        for(int i = 0; i < n; i++) {
            ans[i] = number;
            if(number * 10 <= n) {
                number *= 10;
            } else {
                while(number % 10 == 9 || number + 1 > n) {
                    number /= 10;
                }
                number++;
            }
        }
        return ans;
        
    }
};

File: leetcode_python/386.py
# 386. Lexicographical Numbers
# https://leetcode-cn.com/problems/lexicographical-numbers/
from typing import List
from collections import defaultdict
from collections import deque
from math import inf
import bisect
from functools import lru_cache
import heapq
'''
@File : 386.py
@Time : 2022/04/18 19:52:16
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
'''
class Solution:
    def lexicalOrder(self, n: int) -> List[int]:
        ans = []
        def recursive(val):
            if val > n:
                return
            add = 10
            if val // 10 == 0:
                add = 9 
            for i in range(val, min(n + 1, val + add)):
                ans.append(i)
                recursive(i * 10)
        recursive(1)
        return ans